### PR TITLE
[5.1] Paginator causes superfluous redirect

### PR DESCRIPTION
**Pull request by phroggyy**
Currently, the paginator appends a trailing slash. However, the `.htaccess` currently does 301 redirects to the URL without trailing slash (see https://github.com/laravel/laravel/blob/master/public/.htaccess#L10). Hence, the added trailing slash is seemingly pointless.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Please don't include merge commits.

**Comment by GrahamCampbell** [comment]
Could you add tests please to make sure this behaviour doesn't change again?

**Comment by taylorotwell** [comment]
Why was the trailing slash introduced and when was it introduced?

**Comment by GrahamCampbell** [comment]
I think you introduced it Taylor when you rewrote that stuff?

**Comment by GrahamCampbell** [comment]
See https://github.com/phroggyy/framework/commit/710f64daa212d523c047140b9be963a072a91908.

**Comment by phroggyy** [comment]
That change actually occurred in #b2ee11

**Comment by taylorotwell** [comment]
Gotcha. Can we add a test like Graham said?

**Comment by JosephSilber** [comment]
Not here? https://github.com/laravel/framework/commit/b2ee11148370738ea795ca5c4a36735e7aaa9879

**Comment by phroggyy** [comment]
@JosephSilber see [this line](https://github.com/laravel/framework/commit/b2ee11148370738ea795ca5c4a36735e7aaa9879#diff-9c97dcd5519269329e5390fc31e93f99R99)

**Comment by alexrussell** [comment]
I love how, since Taylor asked, there were at least four of us desperately searching through the repository for when it was introduced.

**Comment by phroggyy** [comment]
@taylorotwell yep, working on it right now

File: tests/Pagination/PaginationPaginatorTest.php [hunk 1]
```diff
@@ -174,4 +174,11 @@ public function testSimplePaginatorReturnsRelevantContextInformation()
             'prev_page_url' => '/?page=1', 'from' => 3, 'to' => 4, 'data' => ['item3', 'item4'],
         ], $p->toArray());
     }
+
+    public function testPaginatorGeneratesUrlsWithoutTrailingSlash()
```
Comment (GrahamCampbell) [inline tests/Pagination/PaginationPaginatorTest.php]: can you check that if you provide a trailing slash in the initial url that it's removed too

Comment (GrahamCampbell) [inline tests/Pagination/PaginationPaginatorTest.php]: for example, to check that `http://website.com/test/` will give `http://website.com/test?page=1` too

**Comment by JosephSilber** [comment]
There are people that strongly oppose using the `page` query variable. These people prefer URLs like this:

> http://example.com/users/page/2

They've badgered Taylor about it, but he didn't see a need for it.

They ended up extending Laravel's paginator to make it work. Can't find it now.

I have a distinct feeling this change will break their extended class.

**Comment by phroggyy** [comment]
@JosephSilber hmm, good point. It depends on what their rewriting does. If it replaces `?` and `=` with a `/`, it wouldn't break, if it replaces `?` with a space and `=` with `/`, it would

**Comment by alexrussell** [comment]
Though hopefully they do it right and generate their own URLs from scratch :)

**Comment by JosephSilber** [comment]
They probably do, but they probably use `$this->path` as the base (I would).

**Comment by alexrussell** [comment]
@JosephSilber yeah good point. And if they look into the code they'd see that a trailing slash is always there, so they'd rely on that too (rather than doing exactly what Taylor did here and `rtrim()` it and then add it, just to be safe).

**Comment by GrahamCampbell** [comment]
Paths should NOT contain query parameters. There is a separate way to add those.

File: tests/Pagination/PaginationPaginatorTest.php (hunk 1 above)
Comment (alexrussell) [inline tests/Pagination/PaginationPaginatorTest.php]: Should also ensure that it works on the base root too (e.g. `http://website.com/`) as that *should* end in a slash, right? I know very few people have pagination on their home page, but something like Tumblr would.

Comment (GrahamCampbell) [inline tests/Pagination/PaginationPaginatorTest.php]: > as that should end in a slash, right?

No. Not according to these discussions.

Comment (phroggyy) [inline tests/Pagination/PaginationPaginatorTest.php]: @alexrussell the existing tests do that

**Comment by taylorotwell** [comment]
So current status of this PR is that we're not sure if it will break an extension that puts the page directly in the URL?

**Comment by phroggyy** [comment]
@taylorotwell well, it will break any extensions with aforementioned implementation, where you do something like `$this->path.'page/'.$page`

**Comment by alexrussell** [comment]
Easily fixable with `rtrim($this->path, '/').'/page/'.$page` - could go in 'upgrade notes' in docs, or maybe just keep an open issue here with the fix clearly stated - I guess it depends on whether people search issues before posting first.

**Comment by phroggyy** [comment]
I agree with @alexrussell on placing it in upgrade notes, unless there are some real downsides to it

**Comment by GrahamCampbell** [comment]
:+1:

**Comment by GrahamCampbell** [comment]
People should never blindly composer update, so their own fault if their app breaks.

**Comment by crynobone** [comment]
@GrahamCampbell as if we have a complete changelog so people can review before running composer update.

**Comment by GrahamCampbell** [comment]
@crynobone We DO document breaking changes though, so people CAN review the upgrading guide before upgrading patch releases.

File: tests/Pagination/PaginationPaginatorTest.php (hunk 1 above)
Comment (alexrussell) [inline tests/Pagination/PaginationPaginatorTest.php]: @phroggyy Ahh sorry, I didn't look at the rest of the test. Also I didn't notice that the line itself checks whether of not the path is the root `/` path anyway, so it's pretty obvious that it *is* accounting for this.

**Comment by taylorotwell** [comment]
While we're on the topic, probably a good idea to keep a real CHANGELOG. Curious how we can get contributors to help maintain that. I always have problems getting people to include their fixes / changes in the log.

Do we agree this is a bug fix or no?

**Comment by GrahamCampbell** [comment]
I'd say it's not really a bugfix, but it's probably a better behaviour.

**Comment by JosephSilber** [comment]
I'm still concerned it'll break things.

People definitely should be able to blindly run composer update to a patch release.

While Laravel doesn't follow semver, it does have its own consistent versioning, and part of that is to not make any real non-BC changes in a patch release.

**Comment by GrahamCampbell** [comment]
> and part of that is to not make any real non-BC changes in a patch release.

That;s not really true. Laravel does make BC breaks in patch releases, just only when it's worth it.

**Comment by taylorotwell** [comment]
Just send this to 5.2. There is no major community outcry for this fix.

**Comment by GrahamCampbell** [comment]
Probably the best idea. ;)

**Comment by garygreen** [comment]
> Just send this to 5.2. There is no major community outcry for this fix.

I've been wanting this for aggges !! Finally after many issues I raised about it at last it's being fixed. Happy to wait until 5.2 though :)

**Comment by alexrussell** [comment]
Just goes to show PRs really do pull more weight than issues.

(Which obviously would make a lot of sense, as it's much easier for maintainers to accept a fix than perform a fix they didn't personally see. Not criticising here!)

**Comment by djtarazona** [comment]
:+1:

I could really use this for 5.1 though!

**Comment by djtarazona** [comment]
On that note, its there any easy way to extend / override this in 5.1? So that `$query->paginate()` returns the correct extended class?  It seems like there is no dependency injection used when a new `LengthAwarePaginator` is created by the `Builder`.

**Comment by garygreen** [comment]
@djtarazona again, I did an issue about that but because there was "no interest" it was closed #7510

**Pull request closed without merging**